src/txrequest.cpp

```diff
@@ -0,0 +1,593 @@
+
+        // This announcement has a predecessor that belongs to the same txhash. Due to ordering, and the
+        // fact that 'it' is not COMPLETED, its predecessor cannot be COMPLETED here.
+        if (it != m_index.get<ByTxHash>().begin() && std::prev(it)->m_txhash == it->m_txhash) return false;
+
+        // This announcement has a successor that belongs to the same txhash, and is not COMPLETED.
+        if (std::next(it) != m_index.get<ByTxHash>().end() && std::next(it)->m_txhash == it->m_txhash &&
+            std::next(it)->m_state != State::COMPLETED) return false;
+
+        return true;
+    }
+
+    /** Convert any announcement to a COMPLETED one. If there are no non-COMPLETED announcements left for this
+     *  txhash, they are deleted. If this was a REQUESTED announcement, and there are other CANDIDATEs left, the
+     *  best one is made CANDIDATE_BEST. Returns whether the announcement still exists. */
+    bool MakeCompleted(Iter<ByTxHash> it)
+    {
+        assert(it != m_index.get<ByTxHash>().end());
+
+        // Nothing to be done if it's already COMPLETED.
+        if (it->m_state == State::COMPLETED) return true;
+
+        if (IsOnlyNonCompleted(it)) {
+            // This is the last non-COMPLETED announcement for this txhash. Delete all.
+            uint256 txhash = it->m_txhash;
+            do {
+                it = Erase<ByTxHash>(it);
+            } while (it != m_index.get<ByTxHash>().end() && it->m_txhash == txhash);
+            return false;
+        }
+
+        // Mark the announcement COMPLETED, and select the next best announcement (the first CANDIDATE_READY) if
+        // needed.
+        ChangeAndReselect(it, State::COMPLETED);
+
+        return true;
+    }
+
+    //! Make the data structure consistent with a given point in time:
+    //! - REQUESTED annoucements with expiry <= now are turned into COMPLETED.
+    //! - CANDIDATE_DELAYED announcements with reqtime <= now are turned into CANDIDATE_{READY,BEST}.
+    //! - CANDIDATE_{READY,BEST} announcements with reqtime > now are turned into CANDIDATE_DELAYED.
+    void SetTimePoint(std::chrono::microseconds now)
+    {
+        // Iterate over all CANDIDATE_DELAYED and REQUESTED from old to new, as long as they're in the past,
+        // and convert them to CANDIDATE_READY and COMPLETED respectively.
+        while (!m_index.empty()) {
+            auto it = m_index.get<ByTime>().begin();
+            if (it->m_state == State::CANDIDATE_DELAYED && it->m_time <= now) {
+                PromoteCandidateReady(m_index.project<ByTxHash>(it));
+            } else if (it->m_state == State::REQUESTED && it->m_time <= now) {
+                MakeCompleted(m_index.project<ByTxHash>(it));
+            } else {
+                break;
+            }
+        }
+
+        while (!m_index.empty()) {
+            // If time went backwards, we may need to demote CANDIDATE_BEST and CANDIDATE_READY announcements back
```

Comment:
Yeah, that works. My earlier argument was that it wasn't much code to deal with time going backwards, and it's elegant that these time-based operations work symmetrically. I've since added a test for time-going-backward behavior, so perhaps it isn't quite true anymore.

I've added a new commit at the end that removes the time-going-backward-makes-candidates-unrequestable-again behavior by introducing a "max time seen" clock inside TxRequestTracker. I think it removes surprisingly little, but it's something.